CompilerSource/parsing/macros.cpp

```diff
@@ -0,0 +1,316 @@
+      if (found == params_by_name.end()) {
+        herr->Error(tokens[i])
+            << "# must be followed by a parameter name; "
+            << tokens[i].content << " is not a parameter";
+        continue;
+      }
+      res.push_back(FuncComponent::Stringify{found->second});
+      e = i + 1;
+    }
+  }
+  if (e < tokens.size())
+    res.push_back(FuncComponent::TokenSpan{e, tokens.size()});
+  return res;
+}
+
+static string TokContentCat(const Token &left, const Token &right) {
+  string content;
+  content.reserve(left.content.size() + right.content.size());
+  content += left.content;
+  content += right.content;
+  return content;
+}
+
+static Token PasteTokens(
+    const Token &left, const Token &right, const ErrorContext &errc) {
+  static const MacroMap no_macros;
+  static const ParseContext empty_context(nullptr);
+  Lexer l(std::move(TokContentCat(left, right)), &empty_context, errc.herr);
+  Token res = l.ReadRawToken();
+  Token end = l.ReadRawToken();
+  if (end.type != TT_ENDOFCODE) {
+    errc.Error() << "Concatenation of `" << left.content
+                 << "` and `" << right.content
+                 << "` does not produce a valid token.";
+  }
+  return res;
+}
+/*
+TokenVector Macro::EvaluateConcats(const TokenVector &replacement_list,
+                                   const ErrorContext &errc) {
+  TokenVector res;
+  res.reserve(replacement_list.size());
+  for (size_t i = 0; i < replacement_list.size(); ++i) {
+    if (replacement_list[i].type == TTM_CONCAT) {
+      const size_t cat_at = i;
+      while (++i < replacement_list.size() &&
+             replacement_list[i].preprocesses_away());
+      while (!res.empty() && res.back().preprocesses_away()) res.pop_back();
+      if (i >= replacement_list.size() || res.empty()) {
+        herr->error(replacement_list[cat_at], kConcatenationError);
+        break;
+      }
+      res.back() = PasteTokens(res.back(), replacement_list[i], errc);
+    } else {
+      res.push_back(replacement_list[i]);
+    }
+  }
+  return res;
+}
+*/
+static void AppendOrPaste(TokenVector &dest,
+                          TokenVector::const_iterator begin,
+                          TokenVector::const_iterator end,
+                          bool paste, const ErrorContext &errc) {
+  if (begin == end) return;
+  if (paste) {
+    while (!dest.empty() && dest.back().PreprocessesAway()) dest.pop_back();
+    while (begin != end && begin->PreprocessesAway()) ++begin;
+    if (!dest.empty() && begin != end) {
+      Token &left = dest.back();
+      left = PasteTokens(left, *begin++, errc);
+    }
+  }
+  dest.insert(dest.end(), begin, end);
+}
+
+TokenVector Macro::SubstituteAndUnroll(
+    const vector<TokenVector> &args, const vector<TokenVector> &args_evald,
+    ErrorContext errc) const {
+  TokenVector res;
+  bool paste_next = false;
+  if (args.size() < parameters->size()) {
+    if (!is_variadic || args.size() + 1 < parameters->size()) {
```

Comment:
im confused at what the difference is between this snipped and the one directly below it